### PR TITLE
BooleanExpression => ConditionExpression with integer values for token count comparisons

### PR DESCRIPTION
**Pull request by atlimit8**
This renames `BooleanExpression` => `ConditionExpression`, changes the underlying value type to `int` and uses the token count as the value of a condition.
This allows the following change
```yaml
RequiresCondition: condition-(n) && !condition-(n+1)
```
=>
```yaml
RequiresCondition: condition == (n)
```
reducing the number of conditions while keeping Boolean operations.

```yaml
GrantCondition@RANK-ELITE:
    RequiresCondition: rank-veteran >= 4
    Condition: rank-elite
```
is used to keep elite ranking decoupled from a specific rank.

### CODE REVIEW

**Comment by reaperrr** [comment]
Travis reports a few style issues.

**Comment by atlimit8** [comment]
yea, I noticed.

**Comment by pchote** [comment]
I was initially on the fence about this idea because eliminating the complexity of upgrade levels was one of my main motivations for introducing conditionals in the first place... but the more I think about this the more I can see this working out well.

The real killer feature will be when we eventually get around to porting prerequisites to conditionals: something like `Prerequisites: (yak + mig) < afld` could be used to restrict the number of aircraft to the number of airfields.

**Comment by atlimit8** [comment]
Note: this is being revised as discussed in the IRC channel.

**Comment by GraionDilach** [comment]
Doesn't this makes StackedConditions redundant?

**Comment by atlimit8** [comment]
Yes, and that will be removed when I'm done updating this PR.

**Comment by GraionDilach** [comment]
In that case someone should add this ticket to the next release milestone.

**Comment by atlimit8** [comment]
Still WIP.
Needs type checking and some operators.

File: OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs [hunk 1]
```diff
@@ -228,14 +196,10 @@ public bool AcceptsExternalCondition(Actor self, string condition, bool timed =
...
-				return true;
-
-			string[] sc;
-			if (stackedConditions.TryGetValue(condition, out sc))
-				return stackedTokens[condition].Count < sc.Length;
+			if (timed)
+				return timers.ContainsKey(condition);
 
-			return conditionCache[condition] == 0;
+			return true;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs]: This method can now simplify down to `return externalConditions.Contains(condition);`

Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs]: Good, I just didn't want to break anything.

File: OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs [hunk 2]
```diff
@@ -219,23 +187,12 @@ public int RevokeCondition(Actor self, int token)
...
+		public bool AcceptsExternalCondition(Actor self, string condition)
 		{
 			if (state == null)
 				throw new InvalidOperationException("AcceptsExternalCondition cannot be queried before the actor has been fully created.");
 
-			if (!externalConditions.Contains(condition))
-				return false;
-
-			// A timed condition can always replace an existing timed condition (resetting its duration)
-			if (timed && timers.ContainsKey(condition))
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs]: Why was this dropped?

Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs]: See outdated review above by @pchote.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs]: This was a workaround for #12498, which is no longer needed because replacing the boolean states with levels implies that whitelisted external conditions are *always* allowed.  I'll have some rather unpleasant words for anyone who suggests that we reintroduce explicit level caps for arbitrary conditions :wink:

File: OpenRA.Game/Support/ConditionExpression.cs [hunk 3]
```diff
@@ -25,47 +25,105 @@ public class ConditionExpression
...
 		{
-			public VariableToken(int index, string symbol)
-				: base(symbol, index, Associativity.Left, 0) { }
+			public static new readonly TokenType Type = new TokenType(0);
+			public VariableToken(int index, string symbol) : base(symbol, index, Type) { }
+		}
+
+		class AndToken : BinaryOperationToken
+		{
+			public static new readonly BinaryOperationTokenType Type = new BinaryOperationTokenType();
```
Comment (pchote) [inline OpenRA.Game/Support/ConditionExpression.cs]: Why are these duplicated everywhere instead of being defined on `BinaryOperationToken` etc?

File: OpenRA.Game/Support/ConditionExpression.cs [hunk 4]
```diff
@@ -30,11 +30,24 @@ class TokenType
 			public readonly int Arity;
 			public readonly int Precedence;
 			public readonly Associativity Associativity;
+			public readonly bool Opener;
```
Comment (pchote) [inline OpenRA.Game/Support/ConditionExpression.cs]: What benefit does this give over type checks?

File: OpenRA.Game/Support/ConditionExpression.cs [hunk 5]
```diff
@@ -64,14 +64,15 @@ class Token
...
+			protected Token(string symbol, int index, TokenType type)
 			{
 				Symbol = symbol;
 				Index = index;
 				Type = type;
 			}
 
-			public bool Is(TokenType type) { return object.ReferenceEquals(Type, type); }
+			public bool Is(TokenType type) { return ReferenceEquals(Type, type); }
+			public virtual void Evaluate(Stack<int> stack, IReadOnlyDictionary<string, int> variables) { }
```
Comment (pchote) [inline OpenRA.Game/Support/ConditionExpression.cs]: These evaluations happen many thousands of times per tick.  Have you checked for performance regressions?

**Comment by pchote** [comment]
General feeling: my simple self contained expression evaluator has turned into an object-oriented monster :sad:

File: OpenRA.Game/Support/ConditionExpression.cs [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+		}
+
+		public enum Keyword { NotAKeyword, True, False, And, Or }
+
+		public static Keyword GetKeyword(string expression)
+		{
+			switch (expression[0])
+			{
+				case 'a':
+					return expression == "and" ? Keyword.And : Keyword.NotAKeyword;
```
Comment (pchote) [inline OpenRA.Game/Support/ConditionExpression.cs]: Please lets not support multiple aliases of the same operator.  It has taken years to force some consistency into our boolean yaml definitions.

File: mods/ts/rules/defaults.yaml [hunk 7]
```diff
@@ -101,7 +107,7 @@
 
 ^Cloakable:
 	Cloak@EXTERNALCLOAK:
-		RequiresCondition: cloakgenerator || crate-cloak
+		RequiresCondition: cloakgenerator + crate-cloak
```
Comment (pchote) [inline mods/ts/rules/defaults.yaml]: I would strongly prefer that we implicitly converted isolated ints to bools in this case and then used the normal boolean operators.  Forcing math operators over boolean operators in this case is a serious step backwards in legibility.

File: mods/ts/rules/shared-vehicles.yaml [hunk 8]
```diff
@@ -129,7 +129,7 @@ LPST:
 	Mobile:
 		Speed: 85
 		TurnSpeed: 5
-		RequiresCondition: !empdisable && undeployed
+		RequiresCondition: !empdisable && ?undeployed
```
Comment (pchote) [inline mods/ts/rules/shared-vehicles.yaml]: I get the impression that this `?` operator and replacing `||` with `+` have come from my request to force like-types around binary operators.  If this is indeed the case, then we should drop that requirement.

File: OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs (hunk 2 above)
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs]: Ah, alright. That also makes following this up with the #12720 solution easier.

File: OpenRA.Game/Support/ConditionExpression.cs (hunk 5 above)
Comment (atlimit8) [inline OpenRA.Game/Support/ConditionExpression.cs]: That was only temporary, I planned on replacing it with `Evaluator` delegate lists when I wrote that commit.

File: mods/ts/rules/shared-vehicles.yaml (hunk 8 above)
Comment (atlimit8) [inline mods/ts/rules/shared-vehicles.yaml]: That is correct. In fact that is the reason I chose Python-like AND and OR operators in the first place. Implicit Boolean evaluation is clearer where the variables are treated both as Boolean values and integers.

File: OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs (hunk 2 above)
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs]: @GraionDilach: If you want to cap effect stacking, make `Targetable` require a condition to be less than / at most a certain amount. That would also give the actors & players feedback by preventing useless attempts.

File: mods/ts/rules/shared-vehicles.yaml (hunk 8 above)
Comment (GraionDilach) [inline mods/ts/rules/shared-vehicles.yaml]: What is that ? operator even supposed to mean?

File: OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs (hunk 2 above)
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Conditions/ConditionManager.cs]: @atlimit8 You can't work around that issue such easily. Such workarounds will result with other gameplay issues.

**Comment by atlimit8** [comment]
I'll be rewriting this with `enum`s, arrays, and fewer `class`es this weekend.

**Comment by pchote** [comment]
Great, thanks @atlimit8.  I'm looking forward to the possibilities of this, so hopefully we can converge on something that everyone is happy without too many more cycles.

File: mods/ts/rules/shared-vehicles.yaml (hunk 8 above)
Comment (atlimit8) [inline mods/ts/rules/shared-vehicles.yaml]: Convert to bool - in order to satisfy strict typing.

**Comment by atlimit8** [comment]
Still WIP with missing operators.

File: OpenRA.Test/OpenRA.Game/ConditionExpressionTest.cs [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+		public void TestParens()
+		{
+			AssertTrue("(true)");
+			AssertTrue("((true))");
+			AssertFalse("(false)");
+			AssertFalse("((false))");
+		}
+
+		[TestCase(TestName = "Arithmetic")]
+		public void TestArithmetic()
```
Comment (pchote) [inline OpenRA.Test/OpenRA.Game/ConditionExpressionTest.cs]: Can you please add some tests using zero and negative numbers?

File: OpenRA.Test/OpenRA.Game/ConditionExpressionTest.cs [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+			AssertValue("5 - 3", 2);
+			AssertValue("2 * 3", 6);
+			AssertValue("6 / 3", 2);
+			AssertValue("7 / 3", 2);
+			AssertValue("8 / 3", 2);
+			AssertValue("6 % 3", 0);
+			AssertValue("7 % 3", 1);
+			AssertValue("8 % 3", 2);
+			AssertValue("7 % 0", 0);
+			AssertValue("8 / 0", int.MaxValue);
```
Comment (pchote) [inline OpenRA.Test/OpenRA.Game/ConditionExpressionTest.cs]: Can you please define / test for the behaviour of `8 / 0 + 1`?

File: OpenRA.Game/Support/ConditionExpression.cs [hunk 11]
```diff
@@ -0,0 +1,901 @@
...
+							continue;
+						}
+
+						case TokenType.Divide:
+						{
+							var y = ast.Pop(ExpressionType.Int);
+							var x = ast.Pop(ExpressionType.Int);
+							var isNotZero = Expressions.Expression.NotEqual(y, Zero);
+							var divide = Expressions.Expression.Divide(x, y);
+							ast.Push(IfThenElse(isNotZero, divide, MaxInt));
```
Comment (pchote) [inline OpenRA.Game/Support/ConditionExpression.cs]: I have two concerns with returning MaxInt here:
* Dividing a negative number by 0 discards the sign
* Integer overflow means that an expresion like `x / total < 10` is true, but `x / total + 1 < 10` is *false* if total == 0.

Returning 0 would be better,  because we're effectively returning NaN as a falsey value.

File: OpenRA.Game/Support/ConditionExpression.cs [hunk 12]
```diff
@@ -0,0 +1,901 @@
...
+					else
+						throw new InvalidProgramException("void BooleanJunction(TokenType type) : type must be And or Or");
+				}
+				else
+				{
+					var y = ast.Pop(ExpressionType.Int);
+					var test = ast.Peek(ExpressionType.Bool);
+					var x = ast.Pop(ExpressionType.Int);
+					if (type == TokenType.And)
+						ast.Push(IfThenElse(test, y, x));
```
Comment (pchote) [inline OpenRA.Game/Support/ConditionExpression.cs]: I don't think python's `and`/`or` behaviour is a good fit here.  It is useful in python as a bodge for a null/falsy-coalescing operator on arbitrary objects, but I find it to be counterintuitive for ints.

Can we please return 1 or 0 instead, for consistency with C/C++?

File: OpenRA.Test/OpenRA.Game/ConditionExpressionTest.cs [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+
+		[TestCase(TestName = "NOT operation")]
+		public void TestNOT()
+		{
+			AssertValue("!true", 0);
+			AssertValue("!false", 1);
+			AssertValue("!!true", 1);
+			AssertValue("!!false", 0);
+			AssertValue("!5", 0);
+			AssertValue("!!5", 1);
```
Comment (pchote) [inline OpenRA.Test/OpenRA.Game/ConditionExpressionTest.cs]: Can you please add tests for `!1`, `!0`, and `!-5`?

File: OpenRA.Test/OpenRA.Game/ConditionExpressionTest.cs [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+			AssertValue("5 < 5", 0);
+			AssertValue("2 <= 5", 1);
+			AssertValue("5 <= 2", 0);
+			AssertValue("5 <= 5", 1);
+			AssertValue("5 > 2", 1);
+			AssertValue("2 > 5", 0);
+			AssertValue("5 > 5", 0);
+			AssertValue("5 >= 2", 1);
+			AssertValue("2 >= 5", 0);
+			AssertValue("5 >= 5", 1);
```
Comment (pchote) [inline OpenRA.Test/OpenRA.Game/ConditionExpressionTest.cs]: Can you please add some tests with zero and negative numbers?

**Review by pchote — COMMENTED** [review]
I've only had a quick skim so far, but this is looking good.  I really like that you are compiling the conditions into Expressions, which is something that I wanted to do the first time around.

Just a few comments to start with, i'll try and have a deeper look later:

File: OpenRA.Game/Support/ConditionExpression.cs [hunk 15]
```diff
@@ -0,0 +1,901 @@
...
+
+						default:
+							throw new InvalidProgramException(
+								"ConditionExpression.Compiler.Compile() is missing an expression builder for TokenType.{0}".F(
+									Enum<TokenType>.GetValues()[(int)t.Type]));
+					}
+				}
+
+				return Expressions.Expression.Lambda<Func<IReadOnlyDictionary<string, int>, int>>(
+					ast.Pop(ExpressionType.Int), SymbolsParam).Compile();
```
Comment (RoosterDragon) [inline OpenRA.Game/Support/ConditionExpression.cs]: Providing a name for the compiled function would be handy for debugging purposes. A fixed string like "ConditionExpression" for a bare minimum of context would be nice. If you wanted to be fancy, maybe try passing the trait name (or some other such contextual strings) down to here so you can give them more detailed names.

File: OpenRA.Test/OpenRA.Game/ConditionExpressionTest.cs (hunk 9 above)
Comment (atlimit8) [inline OpenRA.Test/OpenRA.Game/ConditionExpressionTest.cs]: Done.

File: OpenRA.Mods.Common/Scripting/Properties/ConditionProperties.cs [hunk 16]
```diff
@@ -33,10 +33,10 @@ public ConditionProperties(ScriptContext context, Actor self)
 
 		[Desc("Grant an external condition on this actor and return the revocation token.",
 			"Conditions must be defined on an ExternalConditions trait on the actor.",
-			"If duration > 0 the condition will be automatically revoked after the defined number of ticks")]
+			"duration no longer has any effect and is deprecated")]
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Scripting/Properties/ConditionProperties.cs]: This is wrong - line 42 still has an effect!

Comment (atlimit8) [inline OpenRA.Mods.Common/Scripting/Properties/ConditionProperties.cs]: Shoot, you're right I missed that.

File: OpenRA.Game/Support/ConditionExpression.cs [hunk 17]
```diff
@@ -0,0 +1,898 @@
...
+			public ExpressionType PeekType() { return types[types.Count - 1]; }
+
+			public Expression Peek(ExpressionType toType)
+			{
+				var fromType = types[types.Count - 1];
+				var expression = expressions[expressions.Count - 1];
+				if (toType == fromType)
+					return expression;
+
+				switch (fromType)
```
Comment (pchote) [inline OpenRA.Game/Support/ConditionExpression.cs]: Minor style nit: This would be clearer if you switched on `toType` instead.  I had to stop and read this method carefully when I noticed that the `Bool` case was returning ints and visa versa.

File: OpenRA.Mods.Common/Scripting/Properties/ConditionProperties.cs [hunk 18]
```diff
@@ -48,10 +48,11 @@ public void RevokeCondition(int token)
 			conditionManager.RevokeCondition(Self, token);
 		}
 
-		[Desc("Check whether this actor accepts a specific external condition.")]
+		[Desc("Check whether this actor accepts a specific external condition.",
+		      "The timed parameter does nothing now and is deprecated.")]
```
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Properties/ConditionProperties.cs]: We haven't shipped this method in a release yet, so backwards compatibility isn't a problem.  Please just remove the parameter.

**Review by pchote — COMMENTED** [review]
I am pretty much happy with this now.  Just a couple more minor points:

**Comment by pchote** [comment]
I would quite like to see @chrisforbes and/or @RoosterDragon's opinions on this before we merge.

File: OpenRA.Game/Support/ConditionExpression.cs [hunk 19]
```diff
@@ -0,0 +1,898 @@
...
+		static readonly ParameterExpression SymbolsParam =
+			Expressions.Expression.Parameter(typeof(IReadOnlyDictionary<string, int>), "symbols");
+		static readonly ConstantExpression Zero = Expressions.Expression.Constant(0);
+		static readonly ConstantExpression One = Expressions.Expression.Constant(1);
+		static readonly ConstantExpression False = Expressions.Expression.Constant(false);
+		static readonly ConstantExpression True = Expressions.Expression.Constant(true);
+
+		static Expression AsBool(Expression expression)
+		{
+			return Expressions.Expression.GreaterThan(expression, Zero);
```
Comment (GraionDilach) [inline OpenRA.Game/Support/ConditionExpression.cs]: I'd assume !condition would mean the expression is valid when the condition level is 0 and any other cases I would expect a relation be !d, so yes, I agree.

**Comment by GraionDilach** [comment]
While I welcome the idea behind this PR, it makes me worried that some of the ideas introduced here will be complex for modders to be utilized effectively without a documentation about the introduced expression syntax ideas. That alone shouldn't stand against this PR however. Some of the ideas I can't even think of a usecase at this point even (modulo) but there's been cases where I found such a good while later after the options have been added, so I guess it'll find use somewhere at a point.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 20]
```diff
@@ -466,14 +466,31 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
 							upgradeMaxAcceptedLevel = FieldLoader.GetValue<int>("", maxAcceptedNode.Value.Value);
 
 						var processed = false;
-						if (upgradeTypes.Length == 1 && upgradeMinEnabledLevel == 0 && upgradeMaxEnabledLevel == 0 && upgradeMaxAcceptedLevel == 1)
+						if (upgradeMinEnabledLevel == 0 && upgradeMaxEnabledLevel == 0 && upgradeMaxAcceptedLevel == 1)
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Have you tested that these changed rules work as intended?

File: OpenRA.Mods.Common/Scripting/Properties/ConditionProperties.cs (hunk 18 above)
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Properties/ConditionProperties.cs]: Note that i've made this change as part of #12796.  I suspect that one will be merged first, which cuts a bit of complexity out of this PR.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs (hunk 20 above)
Comment (atlimit8) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Yes:
```
git checkout release-20161019 mods/*/rules/*.yaml
git add mods/*/rules/*.yaml
./OpenRA.Utility.exe ra --upgrade-mod 20161019
git diff
./OpenRA.Utility.exe ts --upgrade-mod 20161019
git diff

```

**Comment by atlimit8** [comment]
@GraionDilach: Yes, documentation will have to be added to the wiki.

**Comment by reaperrr** [comment]
#12796 was merged, needs a rebase.

**Review by chrisforbes — APPROVED** [review]

**Comment by GraionDilach** [comment]
https://github.com/OpenRA/OpenRA/pull/12719/commits/48e5ed49e6f01735df85b8a4872ee636fe848ac9 has a typo in the commit message.

**Review by pchote — APPROVED** [review]
Lets get this merged.  I'll file a followup PR for the negative int to bool case.

**Review by GraionDilach — APPROVED** [review]
LGTM then. :+1:

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/38873e190d89fb0c4bbc8721f3102289965272bc
